CefSharp.Wpf/Internals/MouseAdjustor.cs

```diff
@@ -0,0 +1,132 @@
+using CefSharp.Structs;
+
+namespace CefSharp.Wpf.Internals
+{
+    public class MouseAdjustor
+    {
+        /// <summary>
+        /// The x-offset.
+        /// </summary>
+        private int xOffset;
+
+        /// <summary>
+        /// The y-offset.
+        /// </summary>
+        private int yOffset;
+
+        /// <summary>
+        /// The original-rect.
+        /// </summary>
+        private Rect originalRect;
+
+        /// <summary>
+        /// The teleporting-rect.
+        /// </summary>
+        private Rect teleportingRect;
+
+        public System.Windows.Point Update(Rect originalRect, Rect viewRect)
+        {
+            int x = originalRect.X,
+                prevX = originalRect.X,
+                y = originalRect.Y,
+                prevY = originalRect.Y,
+                xOffset = 0,
+                yOffset = 0;
+
+            // If popup goes outside the view, try to reposition origin
+            if (originalRect.X + originalRect.Width > viewRect.Width)
+            {
+                x = viewRect.Width - originalRect.Width;
+                xOffset = prevX - x;
+            }
+
+            if (originalRect.Y + originalRect.Height > viewRect.Height)
+            {
+                y = y - originalRect.Height - 20;
+                yOffset = prevY - y;
+            }
+
+            // If x or y became negative, move them to 0 again
+            if (x < 0)
+            {
+                x = 0;
+                xOffset = prevX;
+            }
+
+            if (y < 0)
+            {
+                y = 0;
+                yOffset = prevY;
+            }
+
+            if (x != prevX || y != prevY)
+            {
+                this.xOffset = xOffset;
+                this.yOffset = yOffset;
+
+                Rect teleportingRect = new Rect(x, y, x + originalRect.Width, y + originalRect.Height);
```

Comment:
There's still some `teleporting` naming references that should ideally be updated to something nicer

Reply:
Done.